File: backend/src/main/java/practice/project/splitwise/service/BudgetService.java
package practice.project.splitwise.service;

import practice.project.splitwise.dto.BudgetCreationDTO;
import practice.project.splitwise.dto.BudgetDTO;
import practice.project.splitwise.dto.BudgetSummaryDTO;
import practice.project.splitwise.exception.UserNotFoundException;

import java.util.List;

public interface BudgetService {
    
    // Create or update a budget
    BudgetDTO createOrUpdateBudget(BudgetCreationDTO budgetCreationDTO) throws UserNotFoundException;
    
    // Get budget by ID
    BudgetDTO getBudgetById(Integer budgetId);
    
    // Get all budgets for a user in a specific month/year
    List<BudgetDTO> getUserBudgets(Integer userId, Integer month, Integer year);
    
    // Get budget summary for a user in a specific month/year
    BudgetSummaryDTO getBudgetSummary(Integer userId, Integer month, Integer year) throws UserNotFoundException;
    
    // Delete a budget (soft delete)
    void deleteBudget(Integer budgetId);
    
    // Update amount spent for a budget (called when expenses are added)
    void updateAmountSpent(Integer userId, String category, Double amount, Integer month, Integer year);
    
    // Get exceeded budgets for a user
    List<BudgetDTO> getExceededBudgets(Integer userId, Integer month, Integer year);
    
    // Get budgets nearing limit for a user
    List<BudgetDTO> getNearingLimitBudgets(Integer userId, Integer month, Integer year);
} 

File: src/services/budgetService.js
import api from './api';

// Create or update a budget
export const createBudget = (data) => api.post('/api/budget', data);

// Get budget by ID
export const getBudget = (budgetId) => api.get(`/api/budget/${budgetId}`);

// Delete a budget
export const deleteBudget = (budgetId) => api.delete(`/api/budget/${budgetId}`);

// Get all budgets for a user in a specific month/year
export const getUserBudgets = (userId, month, year) => {
  const params = new URLSearchParams();
  if (month) params.append('month', month);
  if (year) params.append('year', year);
  return api.get(`/api/budget/user/${userId}?${params.toString()}`);
};

// Get budget summary for a user
export const getBudgetSummary = (userId, month, year) => {
  const params = new URLSearchParams();
  if (month) params.append('month', month);
  if (year) params.append('year', year);
  return api.get(`/api/budget/summary/${userId}?${params.toString()}`);
};

// Get budgets nearing limit for a user
export const getNearingLimit = (userId, month, year) => {
  const params = new URLSearchParams();
  if (month) params.append('month', month);
  if (year) params.append('year', year);
  return api.get(`/api/budget/nearing-limit/${userId}?${params.toString()}`);
};

// Get exceeded budgets for a user
export const getExceededBudgets = (userId, month, year) => {
  const params = new URLSearchParams();
  if (month) params.append('month', month);
  if (year) params.append('year', year);
  return api.get(`/api/budget/exceeded/${userId}?${params.toString()}`);
}; 